GPU/GX2/TextureCacheGX2.cpp

```diff
@@ -0,0 +1,729 @@
+	case GE_TFMT_5650: return GX2_SURFACE_FORMAT_UNORM_R5_G6_B5;
+	case GE_TFMT_8888:
+	case GE_TFMT_DXT1:
+	case GE_TFMT_DXT3:
+	case GE_TFMT_DXT5:
+	default: return GX2_SURFACE_FORMAT_UNORM_R8_G8_B8_A8;
+	}
+}
+
+TexCacheEntry::TexStatus TextureCacheGX2::CheckAlpha(const u32 *pixelData, u32 dstFmt, int stride, int w, int h) {
+	CheckAlphaResult res;
+	switch (dstFmt) {
+	case GX2_SURFACE_FORMAT_UNORM_R4_G4_B4_A4: res = CheckAlphaRGBA4444Basic(pixelData, stride, w, h); break;
+	case GX2_SURFACE_FORMAT_UNORM_R5_G5_B5_A1: res = CheckAlphaRGBA5551Basic(pixelData, stride, w, h); break;
+	case GX2_SURFACE_FORMAT_UNORM_R5_G6_B5:
+		// Never has any alpha.
+		res = CHECKALPHA_FULL;
+		break;
+	default: res = CheckAlphaRGBA8888Basic(pixelData, stride, w, h); break;
+	}
+
+	return (TexCacheEntry::TexStatus)res;
+}
+
+ReplacedTextureFormat FromGX2Format(u32 fmt) {
+	switch (fmt) {
+	case GX2_SURFACE_FORMAT_UNORM_R5_G6_B5: return ReplacedTextureFormat::F_5650;
+	case GX2_SURFACE_FORMAT_UNORM_R5_G5_B5_A1: return ReplacedTextureFormat::F_5551;
+	case GX2_SURFACE_FORMAT_UNORM_R4_G4_B4_A4: return ReplacedTextureFormat::F_4444;
+	case GX2_SURFACE_FORMAT_UNORM_R8_G8_B8_A8:
+	default: return ReplacedTextureFormat::F_8888;
+	}
+}
+
+GX2SurfaceFormat ToGX2Format(ReplacedTextureFormat fmt) {
+	switch (fmt) {
+	case ReplacedTextureFormat::F_5650: return GX2_SURFACE_FORMAT_UNORM_R5_G6_B5;
+	case ReplacedTextureFormat::F_5551: return GX2_SURFACE_FORMAT_UNORM_R5_G5_B5_A1;
+	case ReplacedTextureFormat::F_4444: return GX2_SURFACE_FORMAT_UNORM_R4_G4_B4_A4;
+	case ReplacedTextureFormat::F_8888:
+	default: return GX2_SURFACE_FORMAT_UNORM_R8_G8_B8_A8;
+	}
+}
+
+void TextureCacheGX2::LoadTextureLevel(TexCacheEntry &entry, ReplacedTexture &replaced, int level, int maxLevel, int scaleFactor, GX2SurfaceFormat dstFmt) {
+	int w = gstate.getTextureWidth(level);
+	int h = gstate.getTextureHeight(level);
+
+	GX2Texture *texture = DxTex(&entry);
+	if ((level == 0 || IsFakeMipmapChange()) && texture == nullptr) {
+		// Create texture
+		int levels = scaleFactor == 1 ? maxLevel + 1 : 1;
+		int tw = w, th = h;
+		GX2SurfaceFormat tfmt = dstFmt;
+		if (replaced.GetSize(level, tw, th)) {
+			tfmt = ToGX2Format(replaced.Format(level));
+		} else {
+			tw *= scaleFactor;
+			th *= scaleFactor;
+			if (scaleFactor > 1) {
+				tfmt = GX2_SURFACE_FORMAT_UNORM_R8_G8_B8_A8;
+			}
+		}
+
+		texture = new GX2Texture();
+		texture->surface.width = tw;
+		texture->surface.height = th;
+		texture->surface.depth = 1;
+		texture->surface.dim = GX2_SURFACE_DIM_TEXTURE_2D;
+		texture->surface.tileMode = GX2_TILE_MODE_LINEAR_ALIGNED;
+		texture->surface.use = GX2_SURFACE_USE_TEXTURE;
+		texture->viewNumSlices = 1;
+		texture->surface.format = tfmt;
+		texture->compMap = GX2_COMP_SEL(_a, _r, _g, _b);
+#if 0 // TODO: mipmapping
+		texture->surface.mipLevels = IsFakeMipmapChange() ? 1 : levels;
+#endif
+
+		GX2CalcSurfaceSizeAndAlignment(&texture->surface);
```

Comment:
Just a note: we make various assumptions in the code that texture memory is aligned to 16 bytes (at least.)  I'm not sure if these assumptions currently affect GX2 code, but I'd suggest rounding the result of this up to 16 during allocation if it can be, e.g., 4.

A note: PSP texture data in PSP RAM/VRAM is always strictly aligned to 16 bytes, including the stride.

-[Unknown]

Reply:
I'm exclusively using the GX2_TILE_MODE_LINEAR_ALIGNED layout for textures, which has much higher alignment requirements for both the buffer(256 bytes) and the stride (128*bpp), so that shouldn't be an issue at all.

Reply:
Ah, cool.

-[Unknown]